### PR TITLE
Graph viz

### PR DESCRIPTION
**Pull request by shaftoes**
Adds preliminary graph visualization functionality. GraphVisualization can create SVG representations of arbitrary lists of courses using GraphViz.

### CODE REVIEW

**Comment by david-yz-liu** [comment]
@shaftoes Just tried recreating the database. Got an error; does this happen to you?
```
Parsing Arts and Science Calendar...
"parsing crs_abs.htm"
Nothing
Just "ABS201Y1"
Nothing
Nothing
Nothing
Nothing
Just "ABS230H1"
Just "ABS201Y1"
Just "ABS201Y1"
Just "ABS201Y1"
database.hs: user error (Text.Regex.Posix.String died: (ReturnCode 8,"parentheses not balanced"))
```

**Comment by shaftoes** [comment]
Mine works fine. I suspect this is due to the different implementations of Regex.Posix between Windows and Linux.

**Comment by christinem** [comment]
If this helps, it works fine on my Mac. It might be a Windows thing? I can try it on my Windows machine to confirm if you like.

**Comment by david-yz-liu** [comment]
@christinem That would be helpful. I will investigate in a bit. 

Thanks, @shaftoes

File: hs/WebParsing/PrerequisiteParsing.hs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+matchDelim prereqs = 
+  let pat = "[;,]" :: String
+  in prereqs =~ pat
+      
+-- | returns true if the string begins inside a parenthesized expression.
+-- e.g  isntDelim "CSC458)" == True
+--      isntDelim "(STA247, STA248)" == False
+isntDelim :: String -> Bool
+isntDelim rest = 
+  let pat = "^.*)" :: String
```
Comment (david-yz-liu) [inline hs/WebParsing/PrerequisiteParsing.hs]: Okay... @shaftoes I'm not entirely sure why this check is necessary, but this is what fails for me - the complaint is about the unmatched parenthesis in this expression. It "works" If I use the string `"^.*\\)"` instead. Can you try this and see if it's fine for you?

**Comment by christinem** [comment]
@david-yz-liu hmmm, seems like the packages I need to run this are not cooperating and installing properly on my Windows Machine.

However, if this helps, I tried changing the string you mentioned above and I can confirm that it works fine on my Mac.

File: hs/WebParsing/PrerequisiteParsing.hs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+matchDelim prereqs = 
+  let pat = "[;,]" :: String
+  in prereqs =~ pat
+      
+-- | returns true if the string begins inside a parenthesized expression.
+-- e.g  isntDelim "CSC458)" == True
+--      isntDelim "(STA247, STA248)" == False
+isntDelim :: String -> Bool
+isntDelim rest = 
+  let pat = "^.[^(]*\\)" :: String
```
Comment (david-yz-liu) [inline hs/WebParsing/PrerequisiteParsing.hs]: Er, I'm not sure why you made this change - it seems quite different from the original.

**Comment by david-yz-liu** [comment]
@shaftoes Poke.

File: hs/WebParsing/ParsingHelp.hs [hunk 3]
```diff
@@ -167,4 +202,17 @@ parseDistAndBreadth (tags, course) =
       brdth = makeEntry (Just (filter (tagContains ["Breadth"]) tags)) (Just ["Breadth Requirement: "])
   in (tail $ tail tags, course {distribution = dist, breadth = brdth})   
 
-
+{-}
```
Comment (david-yz-liu) [inline hs/WebParsing/ParsingHelp.hs]: @shaftoes This this function still necessary?

**Pull request merged**